### DISCUSSION TITLE
Dilate Image According to Alpha?

### DISCUSSION THREAD

**Discussion by Raildex** [ImageMagick/ImageMagick#6755] [Help] [answered]
### ImageMagick version

Magick.NET.Core 13.2.0

### Operating system, version and so on

Win11, 64 Bite

### Description

How do I dilate an image when I want to use the Alpha as a "mask"?
I want to achieve Padding, similar to this:
![Dilation](https://github.com/ImageMagick/ImageMagick/assets/14938934/ed88d96f-5a56-49b3-97f0-fdc61ff768bf)

I have tried

```csharp
newImage.Morphology(MorphologyMethod.Dilate, Kernel.Disk, Channels.Alpha, 4);
newImage.Morphology(MorphologyMethod.Dilate, Kernel.Disk, Channels.RGB, 4);
```
In the case of `Alpha`, it results in a black border around my image.
In the case of `RGB`, the image itself becomes very blurry.

**Comment by fmw42** [comment]
Please provide your input image alone with its alpha channel.  The image you provided has a constant opaque alpha (no variation)

**Reply by Raildex** [reply]
Here is the image I want to dilate
![0](https://github.com/ImageMagick/ImageMagick/assets/14938934/fbf9e9bb-a3f5-405d-83e3-9051a810a117)

and this is the result after 
`newImage.Morphology(MorphologyMethod.Dilate, Kernel.Disk, Channels.Alpha);`
![0_](https://github.com/ImageMagick/ImageMagick/assets/14938934/12148859-3d08-45b0-85a9-a46c5a4937c9)

And here after I use `Channels.RGB`:
![0_](https://github.com/ImageMagick/ImageMagick/assets/14938934/21e98b81-ebfc-44d0-bf64-5fee72312120)

**Reply by fmw42** [reply]
What is the question?  How do you expect the alpha channel to control the dilation?  Can you show an expected result?

**Reply by Raildex** [reply]
@fmw42 I would expect smth along those lines
![grafik](https://github.com/ImageMagick/ImageMagick/assets/14938934/24d73856-1363-4931-bb8a-a60e86fd8f99)

**Comment by fmw42** [comment, marked as answer]
One can extend the image using distort SRT.  Make the viewport larger to your desired size and offset by half the size difference between the input and your desired size. Note the offset needs to be negative in this case.  So here my input is 250x250 and my desired size is 350x350. So the difference is 100 and my offsets are half, namely, 50.

Input:

![image](https://github.com/ImageMagick/ImageMagick/assets/13458218/cee17a51-3f30-426b-8787-38af521c1510)

`magick image.png -virtual-pixel edge -define distort:viewport=350x350-50-50 +distort SRT 1 result.png`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/96bba272-19d3-4188-ab6c-f6b590463073)

If you have an alpha channel in the original, take it and insert in the center of 350x350 black image. Then apply that new alpha to the result above.

See https://imagemagick.org/Usage/distorts/#distort_viewport

**Reply by Raildex** [reply]
Thank you very much! The Magick.NET Equivalent is 
```csharp
croppedSource.Crop(new MagickGeometry(sourceX, sourceY, width, height));
croppedSource.VirtualPixelMethod = VirtualPixelMethod.Edge;
croppedSource.Distort(DistortMethod.ScaleRotateTranslate, new DistortSettings() { Viewport = new MagickGeometry(-Padding/2,-Padding/2,width + Padding, height + Padding) },0.0f);
newTexture.Draw(new DrawableComposite(new MagickGeometry(destinationX, destinationY, width + Padding, height + Padding), CompositeOperator.Copy, croppedSource));
```

**Comment by snibgo** [comment]
Another possibility (Windows BAT syntax):
```
set SP_COLS=0,0,#000,^
%%[fx:w-1],0,#f00,^
0,%%[fx:h-1],#0f0,^
%%[fx:w-1],%%[fx:h-1],#ff0

magick ^
  %INFILE% +write mpr:INP ^
  -size %%[fx:w*3]x%%[fx:h*3] xc:Black ^
  -delete 0 ^
  -sparse-color Bilinear "%SP_COLS%" ^
  -distort DePolar "-1,0" +repage ^
  -distort SRT "0,0 1,%%[fx:1/3] 0" ^
  +repage ^
  -distort Polar "-1,0" +repage ^
  ( mpr:INP -resize 300%% ) ^
  +swap ^
  -compose Distort -composite ^
  g4.png
```
g4.png is:

![g4](https://github.com/ImageMagick/ImageMagick/assets/33812368/ee2f9293-689a-4179-9052-74a8f7b28986)
